src/Okta.Sdk/Internal/ResourceTypeResolver.cs

```diff
@@ -0,0 +1,46 @@
+﻿// <copyright file="ResourceTypeResolver.cs" company="Okta, Inc">
+// Copyright (c) 2014-2017 Okta, Inc. All rights reserved.
+// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+
+namespace Okta.Sdk.Internal
+{
+    internal static class ResourceTypeResolver
+    {
+        private static readonly (Type Resolver, Type For)[] CachedResolvers =
+            typeof(Resource).GetTypeInfo().Assembly.DefinedTypes
+            .Where(typeInfo =>
+            {
+                if (typeInfo.BaseType == null)
+                {
+                    return false;
+                }
+
+                var baseTypeInfo = typeInfo.BaseType.GetTypeInfo();
+                var inheritsFromAbstractResolver = baseTypeInfo.IsGenericType
+                    && baseTypeInfo.GetGenericTypeDefinition() == typeof(AbstractResourceTypeResolver<>);
+
+                return inheritsFromAbstractResolver;
+            })
+            .Select(typeInfo => (typeInfo.AsType(), typeInfo.BaseType.GenericTypeArguments[0]))
+            .ToArray();
```

Comment:
Is it possible for this chain to produce `null` at this point? Looks like when `typeInfo.BaseType == null`, we return `false`, which gets passed into a failing `.Select()` method.

Reply:
Returning `false` from inside the `Where` will simply skip, so that item never makes it to `Select`.

`typeInfo` should never be null, but it could be guarded against by changing the if to `typeInfo?.BaseType == null`

Reply:
added Nate's guard suggestion